Replace pipes with tmp file

sgsutcliffe opened the pull request:
This PR is tackling two things:
The benchmark exposed a scalability issue with using bash commands for `append_clusters` and updating the GAS version to fix some issues with `gas call`

## Addressing the Issue #36 

### Problem
When dealing with large database files to include with the `--db_profiles` and `db_clusters` the process `APPEND_CLUSTERS()` was failing due to a 141 error, which appears to be cause by the pipes in the bash script, specificallt in pipe between `get_address` function and the following awk command to get the number of levels of the genomic address.

```bash
    # Check if two files have consistent delimeter splits in the address column
    init_splits=\$(get_address "${initial_clusters}" | awk -F '${params.gm_delimiter}' '{print NF}')
    add_splits=\$(get_address "${additional_clusters}" | awk -F '${params.gm_delimiter}' '{print NF}')
```
I was currently testing 422 Salmonella samples with about 2500 loci and three levels in both the initial and additional clusters when I noticed the bug.

### Solution
Tried removing the `pipefail` for the process but it didn't work. The current temporary solution is to create a tmp file as an intermediate between `get_address` and the `awk` command. 

## Addressing the Issue #38 and Issue #33 

There were a number of issues we had in `gasnomenclature` were coming from `gas` specifically with `gas call`. We are updating the version of the `genomic address service` aka gas to [0.1.4](https://github.com/phac-nml/genomic_address_service/releases/tag/0.1.4) to address these issues. Will need to revert the tests changed in [PR #31](https://github.com/phac-nml/gasnomenclature/pull/31)

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] Make sure your code lints (`nf-core lint`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

apetkau commented on modules/local/append_clusters/main.nf:
```diff
@@ -4,7 +4,7 @@ process APPEND_CLUSTERS {
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/csvtk:0.22.0--h9ee0642_1' :
-        'biocontainers/csvtk:0.22.0--h9ee0642_1' }"
+        'biocontainers/csvtk:0.22.0--0' }"
```
Why did you revert to container with tag `0` instead of `h9ee0642_1`?

sgsutcliffe replied:
A mistake, when I was trying things out in tests to see if it was a `csvtk` version issue

apetkau commented on modules/local/append_clusters/main.nf:
```diff
@@ -18,15 +18,17 @@ process APPEND_CLUSTERS {
     # Check if two files have consistent delimeter splits in the address column
-    init_splits=\$(get_address "${initial_clusters}" | awk -F '${params.gm_delimiter}' '{print NF}')
-    add_splits=\$(get_address "${additional_clusters}" | awk -F '${params.gm_delimiter}' '{print NF}')
+    get_address "${initial_clusters}" > tmp1.txt
```
This is just a small thing, but could you rename `tmp1` and `tmp2` to correspond to the type of data they contain (e.g., initial_clusters and additional_clusters)?

sgsutcliffe replied:
a6d6de9

apetkau commented on modules/local/cluster_file/main.nf:
```diff
@@ -1,6 +1,5 @@
 process CLUSTER_FILE {
     tag "Create cluster file for GAS call"
-    label 'process_single'
```
This was removed since the resource label isn't needed for a process that runs native code am I correct?

sgsutcliffe replied:
A mistake when I was testing things related to the benchmarking. 3850e90

apetkau reviewed (changes requested):
Thanks so much for fixing this Steven 😄 . In addition to in-line comments could you also add a message to `CHANGELOG.md` about this fix?

sgsutcliffe commented:
The tests that stopped working when we updated GAS to 0.1.3, see the [PR](https://github.com/phac-nml/gasnomenclature/pull/31) due to the threshold issue were:

> * 'Test pipeline with single threshold set to 1'
> * 'Test pipeline with threshold set to 1,0'
> * 'Full pipeline hashes and missing data'
> * 'Full pipeline hashes and missing data count missing as differences'
> * 'Full pipeline remove loci with missing data'

When I updated the GAS container to 0.1.4, with the idea of fixing the issue that caused these tests to fail due to the threshold issue, only one test failed:

> * Full pipeline hashes and missing data

Which was not what I was expecting. It is possible it was the other isssue addressed with GAS 0.1.4 which was regarding the `linkage.method`, which appeared to be hard-coded to `single`. I thought a lot of tests would have failed with 0.1.4 due to this switch since the default was `average` in the nextflow.config. It's possible the `linkage.method` was causing an issue with `gasnomenclature`. 

I will look more closely into the tests to make sure they are behaving as we would expect, based on the GAS versions that were being used.

DarianHole commented on modules/local/append_profiles/main.nf:
```diff
@@ -28,6 +28,9 @@ process APPEND_PROFILES {
     ref_headers=\$(get_header "${reference_profiles}")
     add_headers=\$(get_header "${additional_profiles}")
 
+    echo "\$ref_headers"
+    echo "\$add_headers"
+
```
Was this added for testing purposes or is it for something else I'm overlooking?

sgsutcliffe replied:
It was! Thanks! My test changes keep showing up -- I think I messed up the version I commited.

DarianHole commented on modules/local/append_profiles/main.nf:
```diff
@@ -39,14 +42,15 @@ process APPEND_PROFILES {
-    csvtk mutate2 -t -n new_sample_id -e '(\$source == "db" && \$frequency > 1) ? "db_" + \$sample_id : \$sample_id' | \
-    csvtk cut -t -F -f new_sample_id,l* | \
-    csvtk rename -t -f new_sample_id -n sample_id > profiles_ref.tsv
+    csvtk join -t -f sample_id combined_profiles.tsv sample_counts.tsv |     csvtk mutate2 -t -n new_sample_id -e '(\$source == "db" && \$frequency > 1) ? "db_" + \$sample_id : \$sample_id' > tmp.txt
+    csvtk cut -t -f 2-\${n} tmp.txt > tmp2.txt
+    csvtk cut -t -f new_sample_id tmp.txt | csvtk rename -t -f new_sample_id -n sample_id > tmp3.txt
+    paste tmp3.txt tmp2.txt > profiles_ref.tsv
```
Aaron had mentioned it in the previous review; do we want to adjust the temporary file names to match what they contain here as well?

sgsutcliffe replied:
I had https://github.com/phac-nml/gasnomenclature/commit/a6d6de9be8a0ddc055459c447c49dec525099ccf don't know what happened to the changes!

sgsutcliffe replied:
Oh! I forgot that I did a similar thing on `APPEND_PROFILES` I will make those changes. It was a different issue I was having that came up in the benchmark I forgot I had made those changes.

DarianHole replied:
Hah, sounds good then!

DarianHole reviewed (commented):
Looks great, thanks Steven!! Two small questions from me and that is all

DarianHole reviewed (approved):
Looks good to me now, thanks :)

emarinier commented on modules/local/append_clusters/main.nf:
```diff
@@ -18,15 +18,17 @@ process APPEND_CLUSTERS {
     get_address() {
         if [[ "\${1##*.}" == "gz" ]]; then
-            zcat "\$1" | awk 'NR>1 {print \$2}' | head -n 1
+            zcat "\$1" | awk 'NR>1 {print \$2}'
```
Do you think this pipe might cause issues too?

sgsutcliffe replied:
So far, it has not caused issues when running the benchmark

sgsutcliffe replied:
But it could be a future problem

apetkau replied:
I think the issue with pipes before was that `head -n 1` will exit after printing a single line, which for large files could result in `head` exiting before `awk` finishes, which caused the issue (`awk` I think then had an error of a broken pipe).

For `zcat` and `awk`, `awk` will not exit early and will wait until `zcat` returns an end of file marker, so the same issue should not occur here.

emarinier reviewed (commented):
I'm not saying that you necessarily have to make these changes, but I'm curious if a lot of this would be easier and safer as a small Python program? Or maybe even running `exec`, if it's small enough? Then there isn't the risk of code injection, and there should be less fiddly overall?

apetkau reviewed (approved):
This looks great. Thanks so much @sgsutcliffe 😄
